### DISCUSSION TITLE
How do you create a text content with mixed html

### DISCUSSION THREAD

**Discussion by iongion** [vanjs-org/van#21] [Q&A] [answered]
I am trying to migrate JSX

```jsx
<div class="Callout">
  Long-form information about the important content. This text is styled
  as <a href="#core/typography.running-text">"Running text"</a>, so it
  may contain things like headers, links, lists,
  <code class="bp4-code">code</code> etc.
</div>
```

Supposed we have `Callout` function implemented, how is the `Long-form information...` supposed to be created, as it also has an `<a>` and a `<code>` mixed in ?

Is there a type of node / element without tag, much like Fragment in react ?

**Comment by artydev** [comment]
Hello have tried HTML to VanJS [converter](https://vanjs.org/convert) ?

[Demo](https://flems.io/#0=N4IgtglgJlA2CmIBcA2ADAOgKwHYA0IAZhAgM7IDaoAdgIZiJIgYAWALmLCAQMYD21NvEHIQAHigQAbgAIesWqVIBeADog2ENgnUA+VdRkyAarWoApAMoy2LeDICMGAEwBrAEIyAqgEkZhACd6eAB3PlcDMQB6SSl9akjYmWg1EFoAB3S9AyMJaWSoVM1teGzDGWjY+MrpeINuEFJ4BB5NAXImAGYkNBAAXzwaYNEpMwwAK3JeASERJgQ2GXg8NjxqPFhlAHkAI3H4VrwA7b2DtgwAc3g2AAUAvjYHgE90+C3CPBUACmXV9dgAJTKXQ-AD8oK+TTYABUIAw+ABXNhff4AtahGSWa4AgEYWgwL5sNGFYBXRajWBfAHAALXBEBQy2CCkDBsAZQmQUn7UhYyajKJmkDbKaisgDc8AAhMplLAAGRylEYTYy2CgtjKUiEtZ4HgApDwGUi5UK85QZrXeAotGijXLUV8DCEPgBACitB4LC+hOByOWgJxAIGAmooW5wEFGEd6QRpC98CDweUAS+UDREGUht08Aw1D45uhL3goPgSFDIRk0PgAA8-WiEcofngMC2icDCU6FMiHFE0LjnW6PV6fdm8ZlhFAAMIsEhQL4ptsyqCgkLan7AhP6iCEwNohN4WgijF3PjVp7e7Ut3NA3TABYUIVXiAAXWTKIoaGf4L7Rr4oOoSBUAMV7UM+B7KFAfA8AiDCCBgPC0rQQiuggsHIkSYq0mw9KGLAGDCGwAQQPAWqkP2Lrup6F4UH8z43ne1wyJstA0c+0qZiWwIsWwr6llmtAYFCACCbCERAOxIlaqwJph1pGsuq7rOuIKUvutCBkgKbUECMp-quIEYOa6RCt6V4Ju2lKioQXxmTiB4aZZQZ2QiXy0M2LYQE5gxkkg2rab6GA7BA1BzvaXkAPqZhuGAZjmEBOiQQgpspOYIqCMWkJOIZnPAaZojwjb0by-IUFePFina6IVliyJWV2ACyGRfOFOJOhRQ7ehqt4LMA4VlngCJILAeAXJpfQCp8b4GWATUpaydnDdNs1ZjmAYAmKpDsfKiqbRxeG0ukCg8PAADqWheucDbbmR+pfPt8BgHwUhWmil2ZjifR2VZ7VUXNmw5kSQZ4CEjZlUVjGbOcpCwBAx1fGgeAALQOGixxsDRyrCBctjI2BhQpleeEzek3orfNOJihA1mFO9WE4RQz5iryfDKL1Q0DUg25piNY2YXSDJMW1g6-V1wD8lq6yrN6N5KgOlFei1azvWiDjwJ0r0xRgMZxl8fBOWijoIn0ErVukLqLOahC0AisBsMA+JQEgCIrLQFykEgbmkGwSF8e2vXhek9yPOFfVQCsBobBHECAWBfAx59eBBSFSAhMbDRNC0bTUB0IBoEgDgOAAnIX-SDCAdAMKIExTCA-CCARohwmbATkv49xgDI6gYFEoyipM6gGAYddezIwCkOkZh4DILAOFPsRT3CFxT25cj5ssMg7AEfQyMonKsq7pCD9Qw+LA7O8yF8OyKAmO+6BfzRErfe8O5f19Tw-AJHyfNi0AEZLn5BaCaFLjXBQg9Ai7gng+DnOoDIWQQCf2Pu0RYxQEAAKgjBAiIC2BgLQpA6BXx1CoNKAgo+BheSLF3rEL4wB5CKHdp3EAk5aCwFgIiNg6gBg5EYQAGQEBcRGA4O7BSEUhCAAgZC0B2OwmwdhkhgGbt7QQq965wUrDOUgNgayLGZDIL2TwEBQEkZo9QeBuG0BoSwWkhAkCMIAMT8FpFENgLw+AXCCOkFgTwMABARNQagwUBFCFrJwsx5QjDqAAEp+ICdQC4WiQncG4Wibhpi9F8GSIsGaTwVHe2CrIwJmiYauHsHYfE8AAhCiYsFVwVSYZe0Pkk8o-BzQ0LoUoWx6gdjpAACyIxaSQrh4TGEDNMcksJESQBLDYDwDAqhB5NMQQYF+cCAQ7nThaVo4js6iALkgHAPTS5DErkweCSgGh11mGwUQ0ioA5OANwq+PBXDuMRMnDeR1XBim4fwNhARbFQF-q4F0ZgrjfPKM6QQgj6AkCeLY3+EAWHgr6EfVkWg0EPPKMEtgiMWEQAuABY69cAjgqMJCnFpAIAAC9SwyGcLSMApKZATxgIE2xnQsDpGrMio+di4Gj07uUJ5Lz7h+MdjIP+V8vjOCwFgPADgsAoDwAjTAzgUZMpZZIOJtjnBoC5XSvV1YDX6t1VyplIRoC2FsbKw1TLsW4phgSpARKkpMpmn-YKti0CSKRHwHlCRqAbMztsnOuqej9DAuXYYTB6rGC8KYCwlgvBNEqf0IAA)

Regards

**Reply by iongion** [reply]
Sure, but I want it programmatically

**Reply by artydev** [reply]
What do you think of this  [Demo](https://flems.io/#0=N4IgtglgJlA2CmIBcA2ADAOgKwHYA0IAZhAgM7IDaoAdgIZiJIgYAWALmLCAQMYD21NvEHIQAHWoSAPFAgA3AATQAvGJC0ADhrUA+CdID0suXsnVuIUvAQ82EAeSYBmJGhABfPDXqMQc2tQYAFbkvAJCIkwIbArweGx41HiwygDyAEZB8LZ4AE5pmdlsGADm8GwACrl8bDUAnhrwqYR4pMoAFHEJSbAAlMo6nQD8Q+1WbAAqEAx8AK5s7T29ifAA7goAyuW9vRi0MO1sy1DKwGUx-rDtvcC55bO51ApsLBCkGGye4wqXnTfRCmoyheb2SykCbAA3PAAITKZSwABkiMWGBS8NgQzYylIh0SeB4vSQ8Hh4LRyOKUGs5Xgi2WEOUcUCfAwhD4uQAorQeCx2ocBgs4n0dr1PAJqGs-sAQe8WRpZqRefBRWLlLl2lBlhBGQN4BhqHwqRMGvAhvAkBL1hN4AAPQXLWYdOIYF1HAaHVmwWgLACMBjQuzZnO5vP5Oj1mka1CgAGFXrAoO11W74VAhqs8Z1dTskBBDiLlsq8LRwWsFFU+Da6ny8S79f0dMBohRSHg6xAALpqxYUNAdkYB0l8IbUJBUTx16gd4vKKB8HizBiCDA8O7e+AchBLhZHSF3NgPJ6wDDCNi5CDwXGkQPsrk8msUbodhtN8oKFK0R8duGMs0DT9sF25okjotAYOMACCbBnhA6TzLSCTKnudKkmmGZJFmgxXEWtAikg6rUP08LDhmk4YFSGitnydbKu6VyBIQ7Q0TsxZ4fRoosbM7S0G2LoQBxXjnEgeKEQKGDpBA0adIkAkAPo6uGGDanqECsiQQjqpheqzEMSmkDG4pFPAmrLDwHQvgCQIUHWgGQtiTJllsCwMV6bAALKaO0sk7Kyt4hny2KNtEwCyRaeCzEgsB4CU+HuMCrTdmRYCeVpHwsVFSUpSBerCr0kKkD+SIogVv7HncGhejw8AAOoQC8HqOnm15Eu0ZXwGAfByLSyzFI6yocfSvnBveqUpHqRyingqwdDZFlvikxSkLAEBVe0aB4AAtD6yz5Gwj5osIJQvFt04nOqdbHslGh8tlaU7JCECMScjK9Puh4UB2kIAnwpyhVFEV5pq0WxXu9yPO+Q13qGgXAECuJJAkfINqiQZQ159IvcsPrwE4PVKRg8qKu0fADXgLKzO40I2ho7IxFShC0LMsBsMA+xQEgszxLQJSkEgPGkGw67Eu6IWyRo1S1LJoVQPExLJHLEBjtOfBK+4ywSdGSCrJTFhWDYdgOKIaBID6PoAJxmx4XggHQDCiMEoQgPwginqI0w07kFwBAohDVGAChqBgBj+IEIRqPo1DOwLCjAAopAaAEeAKCwPpJ8YSdBmASeSfKCTvrQ6TWEncHQQIScsLkCjuAoyg-AEHzc6QEdRzEbM1wo7TpLQVi9DXOgd9YRx93XgRs533dxLEzO9M3DgxILuTnO3c4LtupTlJu7WngAQnUACSiZqJGah5RHEiELMkcG087CcBMfCb9u7S330McSAoH915XQhgBV67L-ORcp4VxriEI-U87QADkP8-5CEgafJ4n8X7txfh8c8YBriQgUAYAwCgABy8AuqVzeuDWgzxbQxANFSBQfBCAKFWK8IQ8duTwAUN3Z4LBWF3FIEzNg79P4wNcipagEpcgAAkJhuQADIoI4LASE-CP4kKeII9cK5wjAOILkAWcYSBQAUdQdwZ9I5zzju3AABjIeQCgeBelIG0NQMZaCwFgHMPhIBTAfykQIEoG1M5KGoJnb09gngFzcRw1h7taYBBiM7CIxQFATFeKQchdolApIFnUBAUB+HsKkGQiu8BCCqBAAAYn4HcAwbAGh8BKLkTQLA6gYFyJfagklfFCDtLoNQAAlVp7TUnuKkAYWgOgk6kD4EoPhiDkp1BseEWgkkOHtJSctAA1qwzh+x4DaKTstagazWzvjeGwVs-CpD8GobY7uDiQDpA0AAFg2pc+AugXnDJef3coPAMCGGMDocxs9qDRycS41I8xZF3wfluCB14I5j0jL0do-DjDIsQR-FOyKQBiOsK4k+eBFEKFBbAcF0yP4z0IrraktgQmOBAE4H0JsNrGzcJ4bwdsmArnsRYOJrsmDpENHM4A-Cu48DWXUuYmsFDpEqmsgxH9+CuNyEgBQUBaC5DWeyAIZR5U+3CH4+gJA6gqvVRAZxBijFmFKZGGOAdEGivFdUS+7MFCLy7u0AATFgLAeAfRYBQHgdamAPXbV1QnGA7SVUerQBoG0Cho2xvjTGuNCabS6tWNAF4KrvXJt1Z0tgG1nEQBKKOKqLtci6uSovSSKq0BsPmHwC1Ec1LWCgG-RBGaoBZvjVgXN-Cq0lBrQoOtjNaiVvVYO6gG0BWlzACbPthiI5ekLrAdtH9ZDxy9MayS+z4DTtcWKvNFDC3LRLfhYt7Bx3VqneeEo7B52xvTZmlgKr0CPokJaiQ4FZjpEgDEYVHbn1IA9amq9k7XD1rHfw-NJ7i2jlvewD9zaeG-rqtK+YtQngAY-gK3IVJlUGglLqh1ErnUqq1dQHVSGzBolqWUSu2GFDhtkJRjatQNAPrTfwztWanBoAXR-AdQ6R0NqPXaWDZ6ECEGmZ+wIriSjEFbWuhQQmp0ztqHOn0C73BUv1rS0QTgAAcSBLbuGnDbHwog3IADUACq1mAgADF2RgHIO4IAA):

```js
import van from "./van.js"

const { span, h1, div, form, input, label, button, hr } = van.tags

const add = (base) => (elt) => van.add(base, elt)

const target = document.getElementById("app");


function htmlToElement(html) {
    var template = document.createElement('template');
    html = html.trim(); // Never return a text node of whitespace as the result
    template.innerHTML = html;
    return template.content.firstChild;
}


const s = `<div class="Callout">
  Long-form information about the important content. This text is styled
  as <a href="#core/typography.running-text">"Running text"</a>, so it
  may contain things like headers, links, lists,
  <code class="bp4-code">code</code> etc.
</div>`

const CallOut = htmlToElement(s)

add(app)(
  div(
    h1("Hello"),
    CallOut
  )
)
```

**Reply by iongion** [reply]
Sure, that could work, but is it vanjs idiomatic ?

Other ways to create nodes from string source

- `new DOMParser().parseFromString(htmlSource, 'text/html')`
- Using [createContextualFragment](https://developer.mozilla.org/en-US/docs/Web/API/range/createContextualFragment)

**Reply by artydev** [reply]
VanJS is very minimalistic, and so, there are many egde cases where you have to create  or  bring your own tools
in order to keep it thin.

There is no specially VanJS way of doing things.
You can adopt the procedural way, the functional way, the object oriented way, it is up to you.

VanJS doesn't impose you anything regarding the coding way, only some  advises :-)

Regards

**Comment by Tao-VanJS** [comment, marked as answer]
The most straightforward way in VanJS is to use primitive-valued arguments (string, number, etc.) in tag functions to represent Text node. Thus as @artydev mentioned, the code generated with VanJS [converter](https://vanjs.org/convert) is pretty idiomatic.

That said, the solution mentioned in https://github.com/vanjs-org/van/discussions/21#discussioncomment-6010860 also works. VanJS is unopinionated thus you can choose the way you like as long as you have a good understanding of the behavior of the code :-)
